src/EditorFeatures/Core.Wpf/Preview/PreviewFactoryService.cs

```diff
@@ -315,46 +362,51 @@ public Task<object> CreateRemovedDocumentPreviewViewAsync(Document document, dou
-            var oldBuffer = CreateNewBuffer(document, cancellationToken);
+            var oldBuffer = createBuffer(document, cancellationToken);
 
             // Create PreviewWorkspace around the buffer to be displayed in the diff preview
             // so that all IDE services (colorizer, squiggles etc.) light up in this buffer.
-            var leftDocument = document.Project
-                .RemoveDocument(document.Id)
-                .AddDocument(document.Name, oldBuffer.AsTextContainer().CurrentText);
-            var leftWorkspace = new PreviewWorkspace(leftDocument.Project.Solution);
-            leftWorkspace.OpenDocument(leftDocument.Id);
+            var leftDocumentId = DocumentId.CreateNewId(document.Project.Id);
+            var solutionWithRemovedDocument = removeTextDocument(document.Project.Solution, document.Id);
+            var leftSolution = addTextDocument(solutionWithRemovedDocument, leftDocumentId, document.Name, oldBuffer.AsTextContainer().CurrentText);
+            var leftDocument = leftSolution.GetTextDocument(leftDocumentId);
+            var leftWorkspace = new PreviewWorkspace(leftSolution);
+            openTextDocument(leftWorkspace, leftDocument.Id);
 
             return CreateRemovedDocumentPreviewViewCoreAsync(oldBuffer, leftWorkspace, leftDocument, zoomLevel, cancellationToken);
         }
 
-        public Task<object> CreateRemovedAdditionalDocumentPreviewViewAsync(TextDocument document, double zoomLevel, CancellationToken cancellationToken)
+        public Task<object> CreateRemovedDocumentPreviewViewAsync(Document document, double zoomLevel, CancellationToken cancellationToken)
         {
-            cancellationToken.ThrowIfCancellationRequested();
-
-            // Note: We don't use the original buffer that is associated with oldDocument
-            // (and possibly open in the editor) for oldBuffer below. This is because oldBuffer
-            // will be used inside a projection buffer inside our inline diff preview below
-            // and platform's implementation currently has a bug where projection buffers
-            // are being leaked. This leak means that if we use the original buffer that is
-            // currently visible in the editor here, the projection buffer span calculation
-            // would be triggered every time user changes some code in this buffer (even though
-            // the diff view would long have been dismissed by the time user edits the code)
-            // resulting in crashes. Instead we create a new buffer from the same content.
-            // TODO: We could use ITextBufferCloneService instead here to clone the original buffer.
-            var oldBuffer = CreateNewPlainTextBuffer(document, cancellationToken);
+            return CreateRemovedTextDocumentPreviewViewAsync(
+                document, zoomLevel,
+                createBuffer: (textDocument, cancellationToken) => CreateNewBuffer((Document)textDocument, cancellationToken),
```

Comment:
I see this kind of pattern in multiple places so wonder whether we can have some helper extension method that hide this  and caller can simple do something like AddDocument and that one does right thing based on some info given.

Reply:
we might use IDocumentService to make each different type of document to do right add, remove, open, text changes?

Reply:
Yep, @CyrusNajmabadi had the same suggestion in the original PR and we all agree some kind of a shim that hides the underlying document kinds involved would be helpful. It would be out of scope of this PR though.

Reply:
Filed https://github.com/dotnet/roslyn/issues/35863 to track the refactoring work.